### ISSUE TITLE
netdata allocates 170MB memory after startup (without the database)

### ISSUE THREAD

**Issue by 74cmonty** [netdata/netdata#4487]
##### Question summary

I have configured a local netdata registry server.
Everything works as expected.
I was expecting that this server would allocate small amount of memory.
Therefore I put it on a container (LXC) with 200MB.

After starting up this container the memory allocation shows +170MB.
In my opinion this is too much.
In order to reduce the memory allocation I set paramter `history = 60`, but this was no help.

###### OS / Environment

Alpine Linux 3.8

###### Expected results

I have seen in another LXC that netdata registry server allocates only 50MB.
This is a size I would expect.

**Comment by ktsaou** [comment]
The memory allocated by the registry is tracked with these charts:

![image](https://user-images.githubusercontent.com/2662304/47536078-0a6ff100-d8c7-11e8-8a9f-e67b46823b1d.png)

As you can see, the global netdata registry has allocated 350MB, but it is huge: 1M persons and 500k machines.

Could you please your versions of these charts?

**Comment by ktsaou** [comment]
Please also post this chart:

![image](https://user-images.githubusercontent.com/2662304/47536501-0cd34a80-d8c9-11e8-9d28-92e155f38e05.png)

**Comment by 74cmonty** [comment]
Hi,
I have taken the screenshots below directly from registry node.
![2018-10-26_11-55-49](https://user-images.githubusercontent.com/12523591/47559630-4a11f980-d916-11e8-9f2a-8e507d796a49.png)
![2018-10-26_11-54-05](https://user-images.githubusercontent.com/12523591/47559650-54cc8e80-d916-11e8-84d5-59745c51a0e8.png)

**Comment by ktsaou** [comment]
ok, so it is not the registry.

Could you please post the legend at the bottom of the menu? Like this:

![image](https://user-images.githubusercontent.com/2662304/47560098-fa383000-d91f-11e8-877e-b74bfbf020b6.png)

**Comment by 74cmonty** [comment]
Sure.
![2018-10-26_12-10-17](https://user-images.githubusercontent.com/12523591/47560351-40899100-d918-11e8-8676-f709cb882ce2.png)

**Comment by ktsaou** [comment]
ok, it is not the db either.

Are you streaming metrics from other servers to it?
Are you pushing statsd metrics to it?

**Comment by 74cmonty** [comment]
Nope.
Actually I don't know how to configure netdata to stream or push data.

This is my config:
````
# global netdata configuration

[global]
        history = 60
[web]
        bind to = 10.97.206.204:19999 192.168.1.204:19999 localhost
[registry]
        enabled = yes
        registry to announce = https://lvenetdata-registry.wdf.sap.corp:443
        registry hostname = LVE - netdata Registry
````

Please note that I have configured a Nginx proxy for user authentication; therefore the registry announce is set to port 443.

**Comment by ktsaou** [comment]
hm... there must be something...

Can you take a snapshot of the dashboard and send it to costa@tsaousis.gr ?

Not a screenshot, but a snapshot. Press the arrow that points up, at the navigation bar of the dashboard.

**Comment by 74cmonty** [comment]
Email sent.

**Comment by ktsaou** [comment]
Received it.
I don't see anything unusual though.

For sure something is allocating a lot of memory on this netdata instance. Can you help me pin point the module?

Edit `netdata.conf` and find this section:

```
[plugins]
	# checks = no
	# nfacct = yes
	# proc = yes
	# diskspace = yes
	# cgroups = yes
	# tc = yes
	# idlejitter = yes
	# freeipmi = yes
	# node.d = yes
	# python.d = yes
	# fping = yes
	# charts.d = yes
	# apps = yes
```

Please uncomment and disable `proc`, then restart netdata. The dashboard will have a lot less charts, but please check what `apps.plugin` reports for netdata at the applications section. Is it still using that much memory?

I guess it will not. This would mean a `proc` module does that.
To find the `proc` module, edit `netdata.conf`, enable again `proc` and now find this section:

```
[plugin:proc]
	# netdata server resources = yes
	# /proc/stat = yes
	# /proc/uptime = yes
	# /proc/loadavg = yes
	# /proc/sys/kernel/random/entropy_avail = yes
	# /proc/interrupts = yes
	# /proc/softirqs = yes
	# /proc/vmstat = yes
	# /proc/meminfo = yes
	# /sys/kernel/mm/ksm = yes
	# /sys/devices/system/edac/mc = yes
	# /sys/devices/system/node = yes
	# /proc/net/dev = yes
	# /proc/net/sockstat = yes
	# /proc/net/sockstat6 = yes
	# /proc/net/netstat = yes
	# /proc/net/snmp = yes
	# /proc/net/snmp6 = yes
	# /proc/net/sctp/snmp = yes
	# /proc/net/softnet_stat = yes
	# /proc/net/ip_vs/stats = yes
	# /proc/net/stat/conntrack = yes
	# /proc/net/stat/synproxy = yes
	# /proc/diskstats = yes
	# /proc/net/rpc/nfsd = yes
	# /proc/net/rpc/nfs = yes
	# /proc/spl/kstat/zfs/arcstats = yes
	# /sys/fs/btrfs = yes
	# ipc = yes
```

I would do a binary search. So disable the first half only, restart netdata and check the memory. If apps.plugin now reports a lot less netdata memory the problem is in the first half (the one you disabled). If the memory is still high, it is at the end half. Repeat that process until you find the offending module.

btw, please don't copy the configs for above. Your server may have a different list. You can find your version of this config using http://your.netdata.ip:19999/netdata.conf

**Comment by ktsaou** [comment]
And if `proc` is not the offending module, don't bother with `[plugin:proc]`. You will have to find the module that causes this. So, do a binary search on the plugins (but don't disable `apps` - you need it to see the memory used).

**Comment by 74cmonty** [comment]
Hi Costa,
I executed the following steps:
1. stop netdata
2. change parameter 
````
[plugins]
        # PATH environment variable = /bin:/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:/sbin:/usr/sbin:/usr/local/bin:/usr/local/sbin
        # PYTHONPATH environment variable =
        proc = no
[...]
````
3. restart netdata

The memory allocation dropped to ~48MB!
And the allocation of `apps.plugin` remains unchanged; please check attached screenshot.
![2018-10-29_10-44-58](https://user-images.githubusercontent.com/12523591/47642177-407fd000-db68-11e8-9581-7f5ba47ce16c.png)

I will now continue with the binary search for the `proc` module.

**Comment by ktsaou** [comment]
ok, so it is one or more modules of `proc`.

I bet it is one or more of `/proc/interrupts`, `/proc/softirqs` and `/proc/net/softnet_stat`. These multiply the metrics collected with the number of CPU cores. And you have 58 cores there.

**Comment by 74cmonty** [comment]
Good bet
;-)

It's `/proc/interrupts` only.
With this configuration the memory allocation is ~ 65MB.
````
[plugin:proc]
        # netdata server resources = no
        # /proc/stat = no
        # /proc/uptime = no
        # /proc/loadavg = no
        # /proc/sys/kernel/random/entropy_avail = no
        /proc/interrupts = no
        # /proc/softirqs = no
        # /proc/vmstat = no
        # /proc/meminfo = no
        # /sys/kernel/mm/ksm = no
        # /sys/devices/system/edac/mc = no
        # /sys/devices/system/node = no
        # /proc/net/dev = no
        # /proc/net/sockstat = no
        # /proc/net/sockstat6 = no
        # /proc/net/netstat = no
        # /proc/net/snmp = no
````

Some additional information that might be relevant:
The Netdata Registry is running in a container (LXC) with setting "unprivileged container":
`These kind of containers use a new kernel feature called user namespaces. All of the UIDs (user id) and GIDs (group id) are mapped to a different number range than on the host machine, usually root (uid 0) became uid 100000, 1 will be 100001 and so on.`
Check [here](https://pve.proxmox.com/wiki/Unprivileged_LXC_containers) for more details.

**Comment by 74cmonty** [comment]
I think one should review the enabled metrics when netdata is running in a container, e.g. LXC.
On one hand side it makes sense to run a local registry in a container, on the other hand side not all metrics must be enabled for this purpose.

**Comment by 74cmonty** [comment]
With the setting
[plugin:proc:/proc/interrupts]
	interrupts per core = no
the memory allocation is back to 140MB.

Please check the screenshot for details and information of metrics collected.
![2018-10-29_11-34-55](https://user-images.githubusercontent.com/12523591/47644588-de769900-db6e-11e8-9552-7c9008cf47c0.png)

**Comment by 74cmonty** [comment]
The metrics collected have dropped to 1370.
![2018-10-29_11-44-58](https://user-images.githubusercontent.com/12523591/47645057-2e099480-db70-11e8-8674-fc15869cb95a.png)

I will configure 
```
[plugin:proc:/proc/interrupts]
        interrupts per core = no

[plugin:proc:/proc/softirqs]
        interrupts per core = no
```

on this container.

**Comment by ktsaou** [comment]
hm... please post the output of this:

```sh
cat /proc/interrupts  | wc -w -m
```

**Comment by 74cmonty** [comment]
Sure.
In container
```
~ # cat /proc/interrupts  | wc -w -m
    13086    144401
~ # exit

```
and on host
```
root@ld4465:~# cat /proc/interrupts  | wc -w -m
  13086  144401

```

**Comment by ktsaou** [comment]
hm... it doesn't make sense...

Please also post this:

```
cat /proc/interrupts  | wc -l
```

**Comment by 74cmonty** [comment]
In container
```
~ # cat /proc/interrupts  | wc -l
220
~ # exit
```
and on host
```
root@ld4465:~# cat /proc/interrupts  | wc -l
220
```

**Comment by ktsaou** [comment]
Well, it really does not make sense now...
Could you please send me a copy of your `/proc/interrupts`? I want to make a test with it, to see how it affects netdata memory. Send it at costa@tsaousis.gr

**Comment by ktsaou** [comment]
received it thanks.

**Comment by 74cmonty** [comment]
Just let me know if you need additional data.
(I'm working with other hosts that run with +100 CPUs and +4TB RAM; maybe this could give you additional input for usage of Netdata.)

**Comment by ktsaou** [comment]
ok, I confirm that with your `/proc/interrupts` netdata needs about 50MB. It does not need that memory when `interrupts per core = no`.

So, it is something related to per core stats.

I need to trace the reason for this.

**Comment by ktsaou** [comment]
In PR #4519 I made per core handling a bit smarter. Now it only enables per core monitoring of the interrupts that are actually used by each core.

This lowered the number of dimensions to 3500 from 14400 and lowered also the amount of memory needed by netdata.

I understand netdata allocates some memory per metric.  It should not be that high (50MB for 10000 metrics is about 5MB per metric. This is way too much and needs to investigated further.

@vlvkobal 	I need your help on this. Can you track down the memory allocated per metric, apart the database?

**Comment by ktsaou** [comment]
So, we merged a fix (#4519)  to prevent netdata from enabling interrupts that are not used.

We also identified (#4524) that netdata needs about 3kb per metric for metadata. This is too big and when multiplied by tens of thousands of metrics to gives a big number.

We opened issue #4529 to examine this in more detail and decide on the action plan to solve it.

So, I am closing this.

Thank you very much for reporting it!
